src/module.c

```diff
@@ -7488,7 +7723,24 @@ int moduleTryServeClientBlockedOnKey(client *c, robj *key) {
 RedisModuleBlockedClient *RM_BlockClient(RedisModuleCtx *ctx, RedisModuleCmdFunc reply_callback,
                                          RedisModuleCmdFunc timeout_callback, void (*free_privdata)(RedisModuleCtx*,void*),
                                          long long timeout_ms) {
-    return moduleBlockClient(ctx,reply_callback,timeout_callback,free_privdata,timeout_ms, NULL,0,NULL,0);
+    return moduleBlockClient(ctx,reply_callback,NULL,timeout_callback,free_privdata,timeout_ms, NULL,0,NULL,0);
+}
+
+/* Block the current client for custom authentication in the background. If custom auth is not in
+ * progress, the API returns NULL. Otherwise, the client is blocked and the RM_BlockedClient is
+ * returned similar to the RM_BlockClient API.
+ * Note: Only use this API from the context of a custom auth callback. */
+RedisModuleBlockedClient *RM_BlockClientOnAuth(RedisModuleCtx *ctx, RedisModuleCustomAuthCallback reply_callback,
```

Comment:
Did you consider making it possible to pass `privdata` here as well?

The original `RM_BlockClient` also doesn't provide a way to do that until the client is unblocked, but this is suboptimal and has proved to be a limitation in the past.

Reply:
From the examples I considered, the only use case of private data is when we unblock the client from a background thread and have some data (like a Auth Result) and want it to be accessed from the context of a reply callback.

I could not think of a situation where the Module provides data on blocking the client (`RM_BlockClient`) and wants to access this later.

Is there an example you had in mind where a Module wants to provide data while when using `RM_BlockClient `? / Is there an example of how this would work?

Reply:
how about a case were the module initiated some request to a remote provider and wants to keep the request ID, or the connection on which it issued it?

Reply:
This still doesn't make much sense to me. What exactly are we attaching the private data to? Who is reading it? If we sent it to a remote provider, that background job needs the information. I think it would make more sense if Redis was tracking the request and unblocking it, so it could re-add the private data.

Reply:
not sure i understand the question, shouldn't the private data be attached to the blocked connection?
my thinking is that the module needs to associate each request to the remote provide with a specific client connection.
i suppose it'll be possible to track the client ID, but the module may need to create some dictionary mapping client ID to each remote provider ID, so letting it store that in the client can make it easier.

Reply:
If we want to attach information to a client from a Module, it seems we could benefit from a more generic solution (not just related to the Blocking Client on Authentication Module API). This can help avoid having a dictionary mapping inside the Module for client ID -> private data. 

Each time a Module wants to hold information on the Client Context, it can attach it through a separate API and fetch it.

This would be useful for both blocking and non blocking scenarios

Reply:
it could be nice, but if the module needs that for multiple purposes (one for auth, and another for something else), it'll need t manage some struct with all the info it may need for various tasks.
instead, letting the module store an additional data when registering a callback (each call back), seems like a standard thing to do, and could be easier for the module to handle (and quicker to figure out how)

Reply:
If we feel strongly about it, I would rather have a mechanism that applies to both module blocking mechanisms like using an API like karthik suggested. I'm still not really convinced in the use case, it's not that hard to store a separate map with a usedID -> requestID mapping.

Reply:
note that an API for adding private data for any blocked client is being added in #11568